modules/messages/src/lib.rs

```diff
@@ -507,20 +530,57 @@ decl_module! {
 					debug_assert_eq!(message.key.lane_id, lane_id);
 
 					total_messages += 1;
-					if lane.receive_message::<T::MessageDispatch>(relayer_id.clone(), message.key.nonce, message.data) {
-						valid_messages += 1;
-					}
+					let dispatch_weight = T::MessageDispatch::dispatch_weight(&message);
+					let receival_result = lane.receive_message::<T::MessageDispatch, T::AccountId>(
+						&relayer_id_at_bridged_chain,
+						&relayer_id_at_this_chain,
+						message.key.nonce,
+						message.data,
+					);
+
+					// note that we're returning unspent weight to relayer even if message has been
+					// rejected by the lane. This allows relayers to submit spam transactions with
+					// e.g. the same set of already delivered messages over and over again, without
+					// losing funds for messages dispatch. But keep in mind that relayer pays base
+					// delivery transaction cost anyway. And base cost covers everything except
+					// dispatch, so we have a balance here.
+					let (unspent_weight, refund_pay_dispatch_fee) = match receival_result {
+						ReceivalResult::Dispatched(dispatch_result) => {
+							valid_messages += 1;
+							(dispatch_result.unspent_weight, !dispatch_result.dispatch_fee_paid_during_dispatch)
+						},
+						ReceivalResult::InvalidNonce
+							| ReceivalResult::TooManyUnrewardedRelayers
+							| ReceivalResult::TooManyUnconfirmedMessages => (dispatch_weight, true),
```

Comment:
Why do we assume the message is paid at the source chain here? Wouldn't it be safer to refund less?

Reply:
Weight formula (=> pre-dispatch weight) includes the cost of `Currency::transfer` call (or whatever happens on the integration level when dispatch fee is paid on target chain) for **every message**. Since we are not dispatching the message , this call is not happened => therefore we're deducting it from the final cost